File: services/Services.BackupCreator/Jobs/DropboxSender.cs
using Dropbox.Api;
using Dropbox.Api.Files;
using Microsoft.Extensions.Logging;
using Services.BackupCreator.Config;
using Services.BackupCreator.Secrets;
using System;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Threading.Tasks;

namespace Services.BackupCreator.Jobs
{
    public class DropboxSender
    {
        private readonly ILogger _logger;
        private readonly SecretsConfig _secrets;
        private readonly DropboxConfig _dropboxConfig;

        public DropboxSender(ILogger<BackupCreatorService> logger, SecretsConfig secrets,
            DropboxConfig dropboxConfig)
        {
            _logger = logger;
            _secrets = secrets;
            _dropboxConfig = dropboxConfig;
        }

        public async Task BackupToDropbox(string file)
        {
            DropboxClient client = GetDropboxClient();
            await BackupFile(file, client);
            await RemoveOldBackups(client);
        }

        private async Task RemoveOldBackups(DropboxClient client)
        {
            var filesList = await client.Files.ListFolderAsync(string.Empty);
            var filesCountToGet = filesList.Entries.Count - _dropboxConfig.MaxFiles;
            if (filesCountToGet > 0)
            {
                var filesToRemove = filesList.Entries.OrderBy(f => f.Name)
                    .Take(filesCountToGet);

                foreach (var fileToRemove in filesToRemove)
                {
                    var res = await client.Files.DeleteAsync("/" + fileToRemove.Name);
                    _logger.LogInformation("Removed old backup file: {removedBackupFileName}", res.Name);
                }
            }
        }

        private async Task BackupFile(string file, DropboxClient client)
        {
            var environment = Environment.GetEnvironmentVariable("HOST") ?? Environment.MachineName;

            var backupFileName = $"{environment}_{Path.GetFileName(file)}";

            using (var stream = new FileStream(file, FileMode.Open))
            {
                _logger.LogInformation("Uploading file {file} to Dropbox", file);

                var response = await client.Files
                    .UploadAsync("/" + backupFileName, WriteMode.Overwrite.Instance, body: stream);

                _logger.LogInformation("Uploaded file {file} to Dropbox", file);
            }
        }

        private DropboxClient GetDropboxClient()
        {
            var config = new DropboxClientConfig("Home_Assistant_Back")
            {
                HttpClient = new HttpClient()
                {
                    Timeout = TimeSpan.FromSeconds(_dropboxConfig.TimeoutSeconds)
                }
            };
            return new DropboxClient(_secrets.DropboxToken, config);
        }
    }
}

File: services/SiteParser/create_backup_job/dropbox_sender.py
from logging import Logger

from configs.secrets import Secrets
from configs.dropbox_config import DropboxConfig
import dropbox
import os


class DropboxSender:
    def __init__(self, logger: Logger, secrets_config: Secrets, dropbox_config: DropboxConfig):
        self.__logger = logger
        self.__dropbox_config = dropbox_config
        self.__dbx = dropbox.Dropbox(secrets_config.dropbox_token)

    def backup_to_dropbox(self, file_path):
        file_name = self.__backup_file(file_path)
        self.__remove_old_backups()
        return file_name

    def __backup_file(self, file_path):
        file_name = "/" + os.path.basename(file_path)
        self.__logger.info("Uploading file {file} to Dropbox".format(file=file_name))
        with open(file_path, "rb") as file:
            self.__dbx.files_upload(file.read(), file_name)

        self.__logger.info("Uploaded file {file} to Dropbox".format(file=file_name))
        return file_name

    def __remove_old_backups(self):
        entries = self.__dbx.files_list_folder('').entries
        files_count_to_get = len(entries) - self.__dropbox_config.max_files
        if files_count_to_get > 0:
            files_to_delete = sorted(entries, key=lambda x: x.name)[:files_count_to_get]
            for file in files_to_delete:
                self.__dbx.files_delete_v2("/" + file.name)
                self.__logger.info("Removed old backup file: {removed_backup_file_name}".format(removed_backup_file_name=file.name))
